AI/Nullkiller/Behaviors/BuildingBehavior.cpp

```diff
@@ -49,26 +49,49 @@ Goals::TGoalVec BuildingBehavior::decompose(const Nullkiller * ai) const
 	auto & developmentInfos = ai->buildAnalyzer->getDevelopmentInfo();
 	auto isGoldPressureLow = !ai->buildAnalyzer->isGoldPressureHigh();
 
+	ai->dangerHitMap->updateHitMap();
+
 	for(auto & developmentInfo : developmentInfos)
 	{
-		for(auto & buildingInfo : developmentInfo.toBuild)
+		bool emergencyDefense = false;
+		uint8_t closestThreat = std::numeric_limits<uint8_t>::max();
+		for (auto threat : ai->dangerHitMap->getTownThreats(developmentInfo.town))
+		{
+			closestThreat = std::min(closestThreat, threat.turn);
+		}
+		for (auto& buildingInfo : developmentInfo.toBuild)
 		{
-			if(isGoldPressureLow || buildingInfo.dailyIncome[EGameResID::GOLD] > 0)
+			if (closestThreat <= 1 && developmentInfo.town->fortLevel() < CGTownInstance::EFortLevel::CASTLE && !buildingInfo.notEnoughRes)
 			{
-				if(buildingInfo.notEnoughRes)
+				if (buildingInfo.id == BuildingID::FORT || buildingInfo.id == BuildingID::CITADEL || buildingInfo.id == BuildingID::CASTLE)
 				{
-					if(ai->getLockedResources().canAfford(buildingInfo.buildCost))
-						continue;
-
-					Composition composition;
+					tasks.push_back(sptr(BuildThis(buildingInfo, developmentInfo)));
+					emergencyDefense = true;
+				}
+			}
+		}
```

Comment:
Possibly you can make it a part of DefenceBehavior. Also if you want to emphasize importance of it you may want to build a composition of DefendTown + BuildThis. DefendTown will act as priority booster here.
So all this makes sense only if there is hero or army defending town. Otherwise you just help enemy. Another option in terms of defense is building a dwelling and buying troops depending on enemy strength.
Btw it is also possible to build a composition of DefendTown + ExecuteHeroChain + BuildThis. So that we build Citadel or Castle only when we see a way to actually protect our town.
FORT is not a protection structure for AI now. Only a waste of money. Later possibly if BattleAI can use fort as an obstacle for attacker but this is unlikely in any nearest future.

Reply:
I removed the fort from valid emergencyDefenses. I agree that it doesn't make enough of a difference to warrant building it.
The problem of making it a composition of that is that currently building is processed before everything else. So if the BuildingBehavior doesn't know about the defending also wanting to get a building, it will still fill the tasks with other building-projects.
So the difficult part would be: How to make sure it doesn't build something else instead when the BuildingBehavior doesn't know about it?
I mean it's possible to find a better solution but it's not trivial.

Reply:
Defense will work only in slow evaluation loop while building behavior acts earlier in fast tasks loop. So yes you are right, without removing this fast task loop all together it will not be possible easily